utils/canary/canary.go

```diff
@@ -0,0 +1,142 @@
+
+	if extraReplicaAdded(rolloutSpecReplica, setWeight) {
+		// In the case where the weight of the stable and canary replica counts cannot be divided evenly,
+		// the controller needs to surges by one to account for both replica counts being rounded up.
+		maxSurge = maxSurge + 1
+	}
+	maxReplicaCountAllowed := rolloutSpecReplica + maxSurge
+
+	allRSs := append(oldRSs, newRS)
+	if checkStableRSExists(newRS, stableRS) {
+		allRSs = append(allRSs, stableRS)
+	}
+
+	totalCurrentReplicaCount := replicasetutil.GetReplicaCountForReplicaSets(allRSs)
+	scaleUpCount := maxReplicaCountAllowed - totalCurrentReplicaCount
+
+	desiredStableRSReplicaCount := int32(math.Ceil(float64(rolloutSpecReplica) * (1 - (float64(setWeight) / 100))))
+	desiredNewRSReplicaCount := int32(math.Ceil(float64(rolloutSpecReplica) * (float64(setWeight) / 100)))
+
+	stableRSReplicaCount := int32(0)
+	newRSReplicaCount := *newRS.Spec.Replicas
+	if checkStableRSExists(newRS, stableRS) {
+		stableRSReplicaCount = *stableRS.Spec.Replicas
+	} else {
+		// If there is no stableRS or it is the same as the newRS, then the rollout does not follow the canary steps.
+		// Instead the controller tries to get the newRS to 100% traffic.
+		desiredNewRSReplicaCount = rolloutSpecReplica
+		desiredStableRSReplicaCount = 0
+	}
+
+	if checkStableRSExists(newRS, stableRS) && *stableRS.Spec.Replicas < desiredStableRSReplicaCount {
+		if scaleUpCount > 0 {
+			stableRSReplicaCount = integer.Int32Min(*stableRS.Spec.Replicas+scaleUpCount, desiredStableRSReplicaCount)
+			scaleUpCount = integer.Int32Max(0, scaleUpCount-(desiredStableRSReplicaCount-*stableRS.Spec.Replicas))
+		}
+	}
+
+	if *newRS.Spec.Replicas < desiredNewRSReplicaCount {
+		if scaleUpCount > 0 {
+			newRSReplicaCount = integer.Int32Min(*newRS.Spec.Replicas+scaleUpCount, desiredNewRSReplicaCount)
+		}
+	}
+
+	minAvailableReplicaCount := rolloutSpecReplica - replicasetutil.MaxUnavailable(rollout)
+	totalCurrentAvailableReplicaCount := replicasetutil.GetAvailableReplicaCountForReplicaSets(allRSs)
+	scaleDownCount := totalCurrentAvailableReplicaCount - minAvailableReplicaCount
+	if scaleDownCount <= 0 {
+		// Cannot scale down stableRS or newRS without going below min available replica count
+		return newRSReplicaCount, stableRSReplicaCount, nil
+	}
+
+	totalAvailableOlderReplicaCount := replicasetutil.GetAvailableReplicaCountForReplicaSets(oldRSs)
+
+	if scaleDownCount <= totalAvailableOlderReplicaCount {
+		//Need to scale down older replicas before scaling down the newRS or stableRS.
+		return newRSReplicaCount, stableRSReplicaCount, nil
+	}
+	scaleDownCount = scaleDownCount - totalAvailableOlderReplicaCount
+
+	if *newRS.Spec.Replicas > desiredNewRSReplicaCount {
+		if scaleDownCount > 0 {
+			newRSReplicaCount = integer.Int32Max(*newRS.Spec.Replicas-scaleDownCount, desiredNewRSReplicaCount)
+			scaleDownCount = integer.Int32Max(0, scaleDownCount-(desiredNewRSReplicaCount-*newRS.Spec.Replicas))
```

Comment:
Can you explain what these calculation are?

Reply:
The function is picking between the desiredNewRSReplicaCount or using scaling down as much as possible. It chooses the max between the two because the max enforces that we do not go below the min available replica count to reach the desiredNewRSReplicaCount and we do not scale down below the desiredNewRSReplicaCount. The second calculation determines the number of replicas were used to scale down the newRS replica count. If we scaled down as much as we could, there are no replicas left to scale down otherwise it takes the difference between the desiredNewRSReplicaCount - newRS Replicas count.

Here is another way of writing this logic:
```
if *newRS.Spec.Replicas-scaleDownCount < desiredNewRSReplicaCount {
  newRSReplicaCount = desiredNewRSReplicaCount
  scaleDownCount = scaleDownCount-(desiredNewRSReplicaCount-*newRS.Spec.Replicas)
} else {
  newRSReplicaCount = *newRS.Spec.Replicas-scaleDownCount
  scaleDownCount = 0
}``

I was originally avoiding adding another if statement, but now that I think about it more, the if statements make it easier to read.